source/dofs/dof_handler_policy.cc

```diff
@@ -609,11 +609,37 @@ namespace internal
           vertex_dof_identities (dof_handler.get_fe_collection().size(),
                                  dof_handler.get_fe_collection().size());
 
+          // first identify vertices we want to exclude from working on.
+          // specifically, these are the vertices of artificial and ghost
+          // cells because at the time when we get here, we do not yet
+          // know DoF indices on ghost cells (and we will never know
+          // them for artificial cells). this is, at least the case for
+          // parallel::distributed::Triangulations.
+          //
+          // this means that we will not unify DoF indices between locally
+          // owned cells and ghost cells, and this is different from what
+          // we would do if the triangulation were not split into subdomains.
+          // on the other hand, DoF unification is only an optimization: we
+          // will still record these identities when we compute hanging
+          // node constraints; we just end up with more DoFs than we would
+          // if we unified DoF indices also between locally owned and ghost
+          // cells, but we end up with a simpler algorithm in return.
+          std::vector<bool> include_vertex = dof_handler.get_triangulation().get_used_vertices();
+          if (dynamic_cast<const parallel::distributed::Triangulation<dim,spacedim> *>
+              (&dof_handler.get_triangulation())
+              != nullptr)
+            for (const auto &cell : dof_handler.active_cell_iterators())
+              if (! cell->is_locally_owned())
```

Comment:
while not at all significant, I noticed that you wrote `! cell->is_locally_owned()` here and `cell->is_locally_owned() == false` later :)

Reply:
Hm, good question. If you want me to, I can change this in follow-up patches -- though there appear to be quite a number of occurrences of `if (! cell->is_...)` in this file.

I agree about the lack of consistency. I tend to be more lenient if the function name starts with `is_` or `has_` because then I can read `if (! cell->is_artificial())` as "if not cell is artificial", which makes some grammatical sense. Whereas a statement such as `if (! tria.n_active_cells())` can not be read that way (similarly for `if (! ptr)`) because it implicitly relies on a conversion to boolean that is not spelled out -- and so I prefer `if (tria.n_active_cells() != 0)` (and `if (ptr != null_ptr)`), both of which I can read as a grammatically more or less  complete sentence.